Sources/NIO/PipePair.swift

```diff
@@ -0,0 +1,139 @@
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+extension NIOFileHandle: Selectable {
+}
+
+final class PipePair: SocketProtocol {
+    typealias SocketType = NIOFileHandle
+
+    let inputFD: NIOFileHandle
+    let outputFD: NIOFileHandle
+
+    init(inputFD: NIOFileHandle, outputFD: NIOFileHandle) {
+        self.inputFD = inputFD
+        self.outputFD = outputFD
+        try! inputFD.withUnsafeFileDescriptor {
+            try NIOFileHandle.setNonBlocking(fileDescriptor: $0)
+            try ignoreSIGPIPE(descriptor: $0)
+        }
+        try! outputFD.withUnsafeFileDescriptor {
+            try NIOFileHandle.setNonBlocking(fileDescriptor: $0)
+            try ignoreSIGPIPE(descriptor: $0)
+        }
+    }
+
+    var description: String {
+        return "PipePair { in=\(inputFD), out=\(outputFD) }"
+    }
+
+    func connect(to address: SocketAddress) throws -> Bool {
+        return true
+    }
+
+    func finishConnect() throws {
+        throw ChannelError.operationUnsupported
+    }
+
+    func write(pointer: UnsafeRawBufferPointer) throws -> IOResult<Int> {
+        return try self.outputFD.withUnsafeFileDescriptor { fd in
+            try Posix.write(descriptor: fd, pointer: pointer.baseAddress!, size: pointer.count)
+        }
+    }
+
+    func writev(iovecs: UnsafeBufferPointer<IOVector>) throws -> IOResult<Int> {
+        return try self.outputFD.withUnsafeFileDescriptor { fd in
+            try Posix.writev(descriptor: fd, iovecs: iovecs)
+        }
+    }
+
+    func sendto(pointer: UnsafeRawBufferPointer, destinationPtr: UnsafePointer<sockaddr>, destinationSize: socklen_t) throws -> IOResult<Int> {
+        throw ChannelError.operationUnsupported
+    }
+
+    func read(pointer: UnsafeMutableRawBufferPointer) throws -> IOResult<Int> {
+        return try self.inputFD.withUnsafeFileDescriptor { fd in
+            try Posix.read(descriptor: fd, pointer: pointer.baseAddress!, size: pointer.count)
+        }
+    }
+
+    func recvfrom(pointer: UnsafeMutableRawBufferPointer, storage: inout sockaddr_storage, storageLen: inout socklen_t) throws -> IOResult<Int> {
+        throw ChannelError.operationUnsupported
+    }
+
+    func sendFile(fd: Int32, offset: Int, count: Int) throws -> IOResult<Int> {
+        throw ChannelError.operationUnsupported
+    }
+
+    func recvmmsg(msgs: UnsafeMutableBufferPointer<MMsgHdr>) throws -> IOResult<Int> {
+        throw ChannelError.operationUnsupported
+    }
+
+    func sendmmsg(msgs: UnsafeMutableBufferPointer<MMsgHdr>) throws -> IOResult<Int> {
+        throw ChannelError.operationUnsupported
+    }
+
+    func shutdown(how: Shutdown) throws {
+        switch how {
+        case .RD:
+            throw ChannelError.operationUnsupported
```

Comment:
Is this a reasonable semantic? Why not close the input FD? Pipes are nearly the only place where `shutdown(.RD)` actually has some remotely observable behaviour.

Reply:
fair.